src/wh_server_crypto.c

```diff
@@ -79,6 +103,244 @@ static int wh_Server_HandleSharedSecretCurve25519(whServerContext* ctx,
+                ret = wh_Crypto_SerializeRsaKey(rsa, max_size, out, &der_size);
+                if (ret == 0) {
+                    packet->pkRsakgRes.keyId = 0;
+                    packet->pkRsakgRes.len = der_size;
+                }
+            } else {
+                /* Must import the key into the cache and return keyid */
+                if (WH_KEYID_ISERASED(key_id)) {
+                    /* Generate a new id */
+                    ret = hsmGetUniqueId(server, &key_id);
+#ifdef DEBUG_CRYPTOCB_VERBOSE
+                    printf("[server] RsaKeyGen UniqueId: keyId:%u, ret:%d\n", key_id, ret);
+#endif
+                }
+
+                ret = wh_Server_CacheImportRsaKey(server, rsa,
+                        key_id, flags, label_size, label);
+#ifdef DEBUG_CRYPTOCB_VERBOSE
+                printf("[server] RsaKeyGen CacheKeyRsa: keyId:%u, ret:%d\n", key_id, ret);
+#endif
+                packet->pkRsakgRes.keyId = (key_id & WH_KEYID_MASK);
+                packet->pkRsakgRes.len = 0;
+            }
+        }
+        wc_FreeRsaKey(rsa);
+    }
+
+    if (ret == 0) {
+        /* set the assigned id */
+        *out_size = WH_PACKET_STUB_SIZE +
+                sizeof(packet->pkRsakgRes) +
+                packet->pkRsakgRes.len;
+    }
+    return ret;
+}
+#endif /* WOLFSSL_KEY_GEN */
+
+static int wh_Server_HandleRsaFunction(whServerContext* server, whPacket* packet,
+    uint16_t *out_size)
+{
+    int ret;
+    RsaKey rsa[1] = {0};
+
+    int op_type     = (int)(packet->pkRsaReq.opType);
+    whKeyId key_id  = WH_MAKE_KEYID(WH_KEYTYPE_CRYPTO,
+                        server->comm->client_id,
+                        packet->pkRsaReq.keyId);
+    word32 in_len   = (word32)(packet->pkRsaReq.inLen);
+    word32 out_len  = (word32)(packet->pkRsaReq.outLen);
+    /* in and out are after the fixed size fields */
+    byte* in        = (uint8_t*)(&packet->pkRsaReq + 1);
+    byte* out       = (uint8_t*)(&packet->pkRsaRes + 1);
+
+#ifdef DEBUG_CRYPTOCB_VERBOSE
+    printf("[server] HandleRsaFunction opType:%d inLen:%u keyId:%u outLen:%u\n",
+            op_type, in_len, key_id, out_len);
+#endif
+    switch (op_type)
+    {
+    case RSA_PUBLIC_ENCRYPT:
+    case RSA_PUBLIC_DECRYPT:
+    case RSA_PRIVATE_ENCRYPT:
+    case RSA_PRIVATE_DECRYPT:
+        /* Valid op_types */
+        break;
+    default:
+        /* Invalid opType */
+        return BAD_FUNC_ARG;
+    }
+
+    /* init rsa key */
+    ret = wc_InitRsaKey_ex(rsa, NULL, server->crypto->devId);
+    /* load the key from the keystore */
+    if (ret == 0) {
+        ret = wh_Server_CacheExportRsaKey(server, key_id, rsa);
+#ifdef DEBUG_CRYPTOCB_VERBOSE
+        printf("[server] CacheExportRsaKey keyid:%u, ret:%d\n", key_id, ret);
+#endif
+        if (ret == 0) {
+            /* do the rsa operation */
+            ret = wc_RsaFunction(in, in_len, out, &out_len,
+                op_type, rsa, server->crypto->rng);
```

Comment:
hmm I know this was existing code, but do all the implementations of RSA support in-place operations? SP, fast math, etc? A quick skim tells me yes, and docs don't mention, but perhaps worth a double check. Here the input and output pointers overlap, since it is overlaid on the same packet. Clearly not causing issues since we are passing wolfCrypt tests, but hopefully not a fluke.

Reply:
Definitely supported.  Remember that all of the RSA functions use the large integer math libraries and simply export the answer from that at the end of the RSA function.